packages/opencensus-web-instrumentation-zone/src/perf-resource-timing-selector.ts

```diff
@@ -0,0 +1,169 @@
+ * @param xhrUrl
+ * @param span
+ */
+export function getXhrPerfomanceData(
+  xhrUrl: string,
+  span: Span
+): XhrPerformanceResourceTiming | undefined {
+  const filteredPerfEntries = getPerfResourceEntries(xhrUrl, span);
+  const possibleEntries = getPossiblePerfResourceEntries(filteredPerfEntries);
+  const bestEntry = getBestPerfResourceTiming(possibleEntries, span);
+  return bestEntry;
+}
+
+/**
+ * First step for the algorithm. Filter the Performance Resource Timings by the
+ * name (it should match the XHR URL), additionally, the start/end timings of
+ * every performance entry should fit within the span start/end timings.
+ * These filtered performance resource entries are considered as possible
+ * entries associated to the xhr.
+ * Those are possible because there might be more than two entries that pass the
+ * filter.
+ * @param xhrUrl
+ * @param span
+ */
+export function getPerfResourceEntries(
+  xhrUrl: string,
+  span: Span
+): PerformanceResourceTiming[] {
+  return performance
+    .getEntriesByType('resource')
+    .filter(entry =>
+      isPerfEntryPartOfXhr(entry as PerformanceResourceTiming, xhrUrl, span)
+    ) as PerformanceResourceTiming[];
+}
+
+/**
+ * Second step for the 'Performance resource timings selector algorithm'.
+ * As the XHR could cause a CORS pre-flight request, we have to look for
+ * possible performance entries either containing cors preflight timings or not.
+ * A possible entry with cors data is when a resource timing entry does not
+ * overlap timings with other resource timing entry. Also, a possible entry
+ * without cors resource timing is when that resource timing entry is not
+ * 'paired' with any other entry.
+ * Thus, for this step traverse the array of resource entries and for every
+ * entry check if it is a possible performance resource entry.
+ * @param perfEntries
+ */
+export function getPossiblePerfResourceEntries(
+  perfEntries: PerformanceResourceTiming[]
+): XhrPerformanceResourceTiming[] {
+  const possiblePerfEntries = new Array<XhrPerformanceResourceTiming>();
+  const pairedEntries = new Set<PerformanceResourceTiming>();
+  let entryI: PerformanceResourceTiming;
+  let entryJ: PerformanceResourceTiming;
+  // As this part of the algorithm traverses the array twice, although,
+  // this array is not large as the performance resource entries buffer is
+  // cleared when there are no more running XHRs.
+  for (let i = 0; i < perfEntries.length; i++) {
+    entryI = perfEntries[i];
+    // Compare every performance entry with its consecutive perfomance entries.
+    // That way to avoid comparing twice the entries.
+    for (let j = i + 1; j < perfEntries.length; j++) {
+      entryJ = perfEntries[j];
+      if (!overlappingPerfResourceTimings(entryI, entryJ)) {
+        // As the entries are not overlapping, that means those timings
+        // are possible perfomance timings related to the XHR.
+        possiblePerfEntries.push({
+          corsPreFlightRequest: entryI,
```

Comment:
this assumes that perfEntries is sorted (which is valid because performance.getEntriesByType() returns sorted data). But it would be good to clarify this in a comment somewhere, maybe in the @param line? Or perhaps call filteredPerfEntries "filteredSortedPerfEntries"?